feat: add number of faulty commits to the output to the action

Daniele-Tentoni opened the pull request:
Adds the number of errors detected by the action to the outputs to be used later in the workflow.

Adds to README an example of use in tandem with another action.

Daniele-Tentoni committed:
fix(action): add errors output to the action

Daniele-Tentoni committed:
test: add tests using github actions

Daniele-Tentoni committed:
docs: add section to explain how to use the action and her output

Daniele-Tentoni committed:
fix: use [[ instead of test

Daniele-Tentoni committed:
test: use right output from action

Daniele-Tentoni committed:
ci: use missing token

asperan commented on script.sh:
```diff
@@ -0,0 +1,15 @@
+    exit 2
+fi
+
+ERRORS=`git log --grep='^\w+(\((\w|-)+\))?!?: .+$' --invert-grep -E --pretty=format:%s | awk 'END {print NR}'`
```
Usage of legacy backticks: use `"$(...)"` instead.

Daniele-Tentoni replied:
Thank you, I'll use the newer notation

asperan commented on script.sh:
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
+
+# Check if `git` exists in the current env.
+if ! command -v git &> /dev/null; then
+    echo "git could not be found"
+    exit 2
+fi
+
+ERRORS=`git log --grep='^\w+(\((\w|-)+\))?!?: .+$' --invert-grep -E --pretty=format:%s | awk 'END {print NR}'`
+echo "$ERRORS"
+
+if [[ -n $GITHUB_ACTIONS && $GITHUB_ACTIONS == true ]]; then
+    # Print to github_output only in a Github Actions environment.
+    echo "errors=$ERRORS" >> "$GITHUB_OUTPUT"
+fi
```
The whole script can be put in the action `run` block, there is no need of a new file. If you fear half-execution, put everything inside a function and call the function. Also, inside the function you can run `set -euo pipefail` to stop the script at the first error.

Daniele-Tentoni replied:
Using a new file for the script allow us to run it in other env and calling it in other ways. If you dislike it, I'll make the change.

asperan replied:
What other env do you have in mind?

Daniele-Tentoni replied:
Local git hooks first of all, they run a script.

asperan replied:
I can see your point, but I think it is a niche usage of this script (one may just copy the command inside their hook); so I wouldn't use a file.

That said, in this case it doesn't matter really much, it is just a <1Kb file. Do as you think it is logical (just notify me what you have chosen).

Daniele-Tentoni replied:
The repo is your, I'll go with the action run ;)

asperan commented on script.sh:
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
+
+# Check if `git` exists in the current env.
+if ! command -v git &> /dev/null; then
+    echo "git could not be found"
+    exit 2
+fi
```
This check is useless here and now, the action is run in a GitHub Runner, it has `git` available. If not, with `set -e`  it will be captured when trying to execute `git log`.

asperan commented on script.sh:
```diff
@@ -0,0 +1,15 @@
+ERRORS=`git log --grep='^\w+(\((\w|-)+\))?!?: .+$' --invert-grep -E --pretty=format:%s | awk 'END {print NR}'`
+echo "$ERRORS"
+
+if [[ -n $GITHUB_ACTIONS && $GITHUB_ACTIONS == true ]]; then
```
Usage of `[[` is superfluous: you do not need pattern matching, `test` or `[` is enough.

Daniele-Tentoni replied:
Thank you, I'll use the `[` notation

asperan commented on action.yml:
```diff
@@ -7,14 +7,20 @@ inputs:
         token: ${{ inputs.token }}
-    - run: >
-        test "$(git log --grep='^\w+(\((\w|-)+\))?!?: .+$' --invert-grep -E --pretty=format:%s | awk 'END {print NR}')" -eq "0"
+    - id: run-script
+      run: >
+        [[ $(./script.sh) -eq "0" ]]
```
The conditional expression is not needed: just running the script have the same result, as the exit status of the step is set to be the same exit status of the script. Still, I don't see an improvement in having a separate file for the script for such a simple action.

Side note: here you don't need pattern matching (i.e. `[[`) too, so `test` or `[` will suffice.

Daniele-Tentoni replied:
Thank you, I'll use the `test` notation

asperan commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,33 @@
+    runs-on: ubuntu-latest
+    name: Run script test
+    steps:
+      - uses: actions/checkout@v4.1.1
```
Here you are not fetching the entire history: `git log` returns just the last commit.

Daniele-Tentoni replied:
Here we need only the last commit with action last version code. The deep fetch is done inside the action code, where `fetch-depth: 0` is used.

asperan replied:
Ok, it makes sense

asperan commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,33 @@
+    runs-on: ubuntu-latest
+    name: Run actions test
+    steps:
+      - uses: actions/checkout@v4.1.1
```
Also here, `checkout` only fetches the last commit.

Daniele-Tentoni replied:
Same as [previous reply](https://github.com/asperan/check-conventional-commits-action/pull/1#discussion_r1380202605)

asperan reviewed (changes requested):
IMO, the documentation improvement is very good, but the other changes should be modified.

Edit: I also found some commit types not compliant with the changes introduced. However this is not a real problem as I'm likely to squash this PR when the requested changes are performed.

Daniele-Tentoni committed:
refactor: use newer  instead of old

Accept the comment of asperan on https://github.com/asperan/check-conventional-commits-action/pull/1\#discussion_r1379799093

Daniele-Tentoni committed:
refactor: use [ instead [[

Accept the comment of asperan on https://github.com/asperan/check-conventional-commits-action/pull/1\#discussion_r1379806337

Daniele-Tentoni committed:
refactor: use test instead [[ in action run

Accept the comment of asperan in https://github.com/asperan/check-conventional-commits-action/pull/1\#discussion_r1379811360

asperan reviewed (changes requested):
These two changes are OK, there was no response on the other two comments:
* https://github.com/asperan/check-conventional-commits-action/pull/1#discussion_r1379801621
* https://github.com/asperan/check-conventional-commits-action/pull/1#discussion_r1379805061

The second one is not that important, but I'm eager to receive a clarification on the first one.

Daniele-Tentoni committed:
fix: use set -e command to catch errors inside the script

Accept the comment of asperan on https://github.com/asperan/check-conventional-commits-action/pull/1\#discussion_r1379805061

Daniele-Tentoni committed:
fix: move logic inside action instead of script
